File: brain/thalyn_brain/provider/registry.py
"""Built-in provider registry.

Mirrors `src-tauri/src/provider/registry.rs`. v0.3 enables the
Anthropic adapter; the OpenAI-compatible / Ollama / llama.cpp / MLX
slots are placeholders so users see what's coming.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any

from thalyn_brain.provider.anthropic import DEFAULT_MODEL, AnthropicProvider
from thalyn_brain.provider.auth import AuthBackend
from thalyn_brain.provider.base import (
    Capability,
    CapabilityProfile,
    ChatChunk,
    ChatErrorChunk,
    LlmProvider,
    ProviderKind,
    ProviderMeta,
    ProviderNotImplementedError,
    ReliabilityTier,
)
from thalyn_brain.provider.mlx import (
    DEFAULT_MODEL as MLX_DEFAULT_MODEL,
)
from thalyn_brain.provider.mlx import (
    MlxProvider,
)
from thalyn_brain.provider.ollama import (
    DEFAULT_MODEL as OLLAMA_DEFAULT_MODEL,
)
from thalyn_brain.provider.ollama import (
    OllamaProvider,
)


class _PlaceholderProvider:
    """Surfaces in the listing but cannot stream."""

    def __init__(self, id: str, display_name: str, kind: ProviderKind) -> None:
        self._id = id
        self._display_name = display_name
        self._kind = kind
        self._profile = CapabilityProfile(
            max_context_tokens=0,
            supports_tool_use=False,
            tool_use_reliability=ReliabilityTier.UNKNOWN,
            supports_vision=False,
            supports_streaming=False,
            local=kind in {ProviderKind.OLLAMA, ProviderKind.LLAMA_CPP, ProviderKind.MLX},
        )

    @property
    def id(self) -> str:
        return self._id

    @property
    def display_name(self) -> str:
        return self._display_name

    @property
    def capability_profile(self) -> CapabilityProfile:
        return self._profile

    @property
    def default_model(self) -> str:
        return ""

    def supports(self, capability: Capability) -> bool:
        return self._profile.supports(capability)

    def stream_chat(
        self,
        prompt: str,
        *,
        history: list[dict[str, Any]] | None = None,
        system_prompt: str | None = None,
    ) -> AsyncIterator[ChatChunk]:
        del prompt, history, system_prompt
        return _placeholder_stream(self._id)


async def _placeholder_stream(provider_id: str) -> AsyncIterator[ChatChunk]:
    raise ProviderNotImplementedError(f"provider {provider_id} is not implemented in v0.3")
    # The yield below is never reached, but its presence forces Python
    # to treat this function as an async generator so the caller can
    # use it with `async for`. mypy's flow analysis is told to ignore.
    yield ChatErrorChunk(message="unreachable")  # type: ignore[unreachable]


def builtin_providers(*, anthropic_auth: AuthBackend | None = None) -> list[LlmProvider]:
    return [
        AnthropicProvider(auth_backend=anthropic_auth),
        _PlaceholderProvider(
            "openai_compat",
            "OpenAI-compatible endpoint",
            ProviderKind.OPENAI_COMPATIBLE,
        ),
        OllamaProvider(),
        _PlaceholderProvider("llama_cpp", "llama.cpp (local)", ProviderKind.LLAMA_CPP),
        MlxProvider(),
    ]


class ProviderRegistry:
    """Lookup + metadata for the built-in providers."""

    def __init__(self, providers: list[LlmProvider] | None = None) -> None:
        self._providers: dict[str, LlmProvider] = {
            provider.id: provider for provider in (providers or builtin_providers())
        }

    def list_meta(self, *, configured: dict[str, bool] | None = None) -> list[ProviderMeta]:
        configured = configured or {}
        enabled_ids = {"anthropic", "ollama", "mlx"}
        out: list[ProviderMeta] = []
        for provider in self._providers.values():
            kind = _kind_for(provider.id)
            default_model = ""
            if provider.id == "anthropic":
                default_model = DEFAULT_MODEL
            elif provider.id == "ollama":
                default_model = OLLAMA_DEFAULT_MODEL
            elif provider.id == "mlx":
                default_model = MLX_DEFAULT_MODEL
            out.append(
                ProviderMeta(
                    id=provider.id,
                    display_name=provider.display_name,
                    kind=kind,
                    default_model=default_model,
                    capability_profile=provider.capability_profile,
                    configured=configured.get(provider.id, False),
                    enabled=provider.id in enabled_ids,
                )
            )
        return out

    def get(self, provider_id: str) -> LlmProvider:
        try:
            return self._providers[provider_id]
        except KeyError as exc:
            raise ProviderNotImplementedError(f"unknown provider: {provider_id}") from exc


def build_registry(*, anthropic_auth: AuthBackend | None = None) -> ProviderRegistry:
    return ProviderRegistry(builtin_providers(anthropic_auth=anthropic_auth))


def _kind_for(provider_id: str) -> ProviderKind:
    match provider_id:
        case "anthropic":
            return ProviderKind.ANTHROPIC
        case "openai_compat":
            return ProviderKind.OPENAI_COMPATIBLE
        case "ollama":
            return ProviderKind.OLLAMA
        case "llama_cpp":
            return ProviderKind.LLAMA_CPP
        case "mlx":
            return ProviderKind.MLX
        case _:
            return ProviderKind.OPENAI_COMPATIBLE

File: src-tauri/src/provider/registry.rs
//! Built-in provider registry.
//!
//! Ships with a static list of provider metadata. v0.3 enables the
//! Anthropic provider; the OpenAI-compatible / Ollama / llama.cpp /
//! MLX entries surface in the UI as disabled placeholders so users
//! can see what's coming without being able to select them yet.

use std::sync::Arc;

use async_trait::async_trait;

use super::{
    CapabilityProfile, LlmProvider, ProviderError, ProviderKind, ProviderMeta, ReliabilityTier,
};

/// Static metadata for the Claude Sonnet 4.6 default. The capability
/// numbers track public Anthropic documentation as of the v0.3
/// release; they're inputs to the UI delta surface, not API calls.
pub const ANTHROPIC_DEFAULT_MODEL: &str = "claude-sonnet-4-6";

pub struct AnthropicProvider {
    profile: CapabilityProfile,
    configured: bool,
}

impl AnthropicProvider {
    pub fn new(configured: bool) -> Self {
        Self {
            profile: CapabilityProfile {
                max_context_tokens: 200_000,
                supports_tool_use: true,
                tool_use_reliability: ReliabilityTier::High,
                supports_vision: true,
                supports_streaming: true,
                local: false,
            },
            configured,
        }
    }
}

#[async_trait]
impl LlmProvider for AnthropicProvider {
    fn id(&self) -> &str {
        "anthropic"
    }
    fn display_name(&self) -> &str {
        "Anthropic — Claude Sonnet 4.6"
    }
    fn capability_profile(&self) -> &CapabilityProfile {
        &self.profile
    }
    async fn probe(&self) -> Result<(), ProviderError> {
        if self.configured {
            Ok(())
        } else {
            Err(ProviderError::NotConfigured("anthropic".into()))
        }
    }
}

/// A v0.3 placeholder — the UI lists these so the user can see what
/// providers are coming, but selecting them is disabled. The probe
/// always returns `NotImplemented`.
struct PlaceholderProvider {
    id: &'static str,
    display_name: &'static str,
    profile: CapabilityProfile,
}

#[async_trait]
impl LlmProvider for PlaceholderProvider {
    fn id(&self) -> &str {
        self.id
    }
    fn display_name(&self) -> &str {
        self.display_name
    }
    fn capability_profile(&self) -> &CapabilityProfile {
        &self.profile
    }
    async fn probe(&self) -> Result<(), ProviderError> {
        Err(ProviderError::NotImplemented(self.id.to_string()))
    }
}

fn placeholder_profile(reliability: ReliabilityTier, local: bool) -> CapabilityProfile {
    CapabilityProfile {
        max_context_tokens: 0,
        supports_tool_use: false,
        tool_use_reliability: reliability,
        supports_vision: false,
        supports_streaming: false,
        local,
    }
}

/// Build the v0.3 provider list. The configured flag for Anthropic is
/// supplied by the caller so the registry stays decoupled from the
/// keychain.
pub fn builtin_providers(anthropic_configured: bool) -> Vec<Arc<dyn LlmProvider>> {
    vec![
        Arc::new(AnthropicProvider::new(anthropic_configured)),
        Arc::new(PlaceholderProvider {
            id: "openai_compat",
            display_name: "OpenAI-compatible endpoint",
            profile: placeholder_profile(ReliabilityTier::Unknown, false),
        }),
        Arc::new(PlaceholderProvider {
            id: "ollama",
            display_name: "Ollama (local)",
            profile: placeholder_profile(ReliabilityTier::Unknown, true),
        }),
        Arc::new(PlaceholderProvider {
            id: "llama_cpp",
            display_name: "llama.cpp (local)",
            profile: placeholder_profile(ReliabilityTier::Unknown, true),
        }),
        Arc::new(PlaceholderProvider {
            id: "mlx",
            display_name: "MLX (Apple Silicon)",
            profile: placeholder_profile(ReliabilityTier::Unknown, true),
        }),
    ]
}

/// Top-level registry the rest of the runtime addresses. Holds the
/// `Arc`-shared providers and exposes them as wire-friendly metadata
/// for the renderer.
pub struct ProviderRegistry {
    providers: Vec<Arc<dyn LlmProvider>>,
}

impl ProviderRegistry {
    pub fn new(providers: Vec<Arc<dyn LlmProvider>>) -> Self {
        Self { providers }
    }

    pub fn list_meta(&self) -> Vec<ProviderMeta> {
        self.providers
            .iter()
            .map(|provider| {
                let id = provider.id();
                let kind = match id {
                    "anthropic" => ProviderKind::Anthropic,
                    "openai_compat" => ProviderKind::OpenAiCompatible,
                    "ollama" => ProviderKind::Ollama,
                    "llama_cpp" => ProviderKind::LlamaCpp,
                    "mlx" => ProviderKind::Mlx,
                    _ => ProviderKind::OpenAiCompatible,
                };
                let default_model = match kind {
                    ProviderKind::Anthropic => ANTHROPIC_DEFAULT_MODEL.to_string(),
                    _ => String::new(),
                };
                let configured = matches!(kind, ProviderKind::Anthropic)
                    && provider.capability_profile().max_context_tokens > 0;
                let enabled = matches!(kind, ProviderKind::Anthropic);
                ProviderMeta {
                    id: id.to_string(),
                    display_name: provider.display_name().to_string(),
                    kind,
                    default_model,
                    capability_profile: provider.capability_profile().clone(),
                    configured,
                    enabled,
                }
            })
            .collect()
    }

    pub fn find(&self, id: &str) -> Option<Arc<dyn LlmProvider>> {
        self.providers.iter().find(|p| p.id() == id).map(Arc::clone)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::provider::Capability;

    #[tokio::test]
    async fn anthropic_probe_requires_configuration() {
        let provider = AnthropicProvider::new(false);
        assert!(matches!(
            provider.probe().await,
            Err(ProviderError::NotConfigured(_))
        ));
        let provider = AnthropicProvider::new(true);
        assert!(provider.probe().await.is_ok());
    }

    #[test]
    fn anthropic_supports_streaming_and_tools() {
        let provider = AnthropicProvider::new(true);
        assert!(provider.supports(Capability::Streaming));
        assert!(provider.supports(Capability::ToolUse));
        assert!(provider.supports(Capability::Vision));
    }

    #[tokio::test]
    async fn placeholders_are_not_implemented() {
        let registry = ProviderRegistry::new(builtin_providers(false));
        let ollama = registry.find("ollama").expect("ollama present");
        assert!(matches!(
            ollama.probe().await,
            Err(ProviderError::NotImplemented(_))
        ));
    }

    #[test]
    fn registry_marks_anthropic_enabled_others_disabled() {
        let registry = ProviderRegistry::new(builtin_providers(true));
        let metas = registry.list_meta();
        let anthropic = metas
            .iter()
            .find(|m| m.id == "anthropic")
            .expect("anthropic present");
        assert!(anthropic.enabled);
        assert!(anthropic.configured);
        for other in metas.iter().filter(|m| m.id != "anthropic") {
            assert!(!other.enabled, "{} should be disabled in v0.3", other.id);
        }
    }
}
